lib/rucio/rse/rsemanager.py

```diff
@@ -17,6 +17,7 @@
 import logging
 import random
 from time import sleep
+from typing import Dict, List, Union
```

Comment:
Yes, a single commit is preferable.

Reply:
I can try to make a separate commit. It will be good for practice

Reply:
I have made changes in commit history.

Also Python type check fails in `_get_possible_protocols` due to `bool` type when `protocols` type-hinted as `list[dict[str, Union[str, int, dict[str, int]]]]` (where `bool` is not present) is called.

Please, can you say am I something missing or it is not a fault on my side?

Reply:
Note that in the first commit you’re importing `typing.Union` (line 20) and breaking the definition of `upload()` (lines 296–297). These don’t match the commit description ‘spelling fixes’ and should be move to the second commit.

Also note the improper break before a comma in line 70.

As for you question, simple type type hints can only get us so far. Suppose you have a dictionary structure of:
```python
d = {
    'a_bool': True,
    'a_list': ['a', 'b']
}
```
If you type hint it as `dict[str, Union[bool, list]`, you aren’t specifying that `a_bool` is expected to be a `bool` and `a_list` is expected to be a `list[str]`. Hence, if somewhere below you do something like `d['a_list'].append('c')`, it will raise an error because `d['a_list']` is hinted as `Union[bool, list]`, so it *may* be a `bool`.

To be frank, I think it’s actually better to use `dict[str, Any]`. It creates a placeholder for future work: using `typing.TypedDict` to unequivocally define the structure and expected content of dictionaries. We’ve discussed it multiple times, but haven’t started working on it.

Reply:
Thank you for pointing out issues in first commit!

I will change type-hinting to `dict[str, Any]` in second commit. Don't know if type-hinting with `typing.TypedDict` can be done right in parameters of a function but I think it could look something like this: 
```
Protocols = TypedDict('Protocols',
                      {'scheme': str,
                       'hostname': str,
                       'prefix': str,
                       'port': int,
                       'impl': str,
                       'extended_attributes': dict[str, str],
                       'domains': dict[str, int]
                       })

Rse_settings = TypedDict('Rse_settings',
                         {'id': str,
                          'rse': str,
                          'type': str,
                          'volatile': bool,
                          'verify_checksum': bool,
                          'deterministic': bool,
                          'domain': str,
                          'protocols': list[Protocols]
                          })


def _get_possible_protocols(rse_settings: Rse_settings, operation, scheme=None, domain=None, impl=None):
```
or implementation of such syntax will be discussed later for whole project (like migration to SQLAlchemy 2.0)?

Reply:
This is my personal opinion, but I believe it to be a structured task that will involve:

 1. List all functions that have dictionaries as parameters or return values.
 2. Identify the minimal set of types that are common across the entire Rucio project.
 3. Define them as `TypedDicts` is a common file like `lib/rucio/common/types.py`.
 4. Use them, refactoring the code as necessary (some arguments can be multiple, incompatible types, which makes type checking cumbersome).

One potential hurdle is types where the existence of elements is optional. [Python 3.11](https://docs.python.org/3/whatsnew/3.11.html#whatsnew311-pep655) offers support for them, but it will be quite some time before we can raising the minimum required version.

Reply:
I’ve changed my mind, sorry! I believe it would be a waste to not make use of the work you’ve already done. Here are some guidelines on how to proceed:

 1. Definitions should go to `lib/rucio/common/types.py`.
 2. They should use camel case and have the `Dict` suffix.
 3. They should be defined as classes, inheriting from `typing.TypedDict`.
 4. In case there’s any key whose presence is optional, then that line should end with a comment `# FIXME: typing.NotRequired`.

I would like you to review the code at:

https://github.com/rucio/rucio/blob/a73004e1e6ecc5815c2c11a4c80e641be0b97930/lib/rucio/core/rse.py#L1376-L1450

and then define `RSEDomainWANDict`, `RSEDomainLANDict`, `RSEDomainsDict`, `RSEProtocolDict` and `RSESettingsDict`. Then comes the difficult part: you need to start from that function and add hints when necessary, while also reviewing whether the structure is changed.

Does that interest you? I will be happy to assist along the way.

Reply:
Yeah, I will be happy to implement this! 

So to start with, I need to define in `lib/rucio/common/types.py` classes 
`RSEDomainWANDict`, 
`RSEDomainLANDict`, 
`RSEDomainsDict`, 
`RSEProtocolDict` and 
`RSESettingsDict`, 
all with inheritance from `typing.TypedDict`.

Then if key inside one of these classes is optional I write something like:
```
class RSEDomainsDict(typing.TypedDict):
    name: str # FIXME: typing.NotRequired
```
After that I start type-hinting when necessary in `def _format_get_rse_protocols()` using early defined classes for hints while also checking if the dictionaries I have defined can represent all of the possible data structures that can be returned by the function. 

Am I understanding you correctly?

Reply:
Thank you for the enthusiasm. Yes, that sounds about right. For `_format_get_rse_protocols()` you’ll only need to replace the return type with `RSESettingsDict`, then do the same for `get_rse_protocols()`, then follow until you eventually get back to the RSE Manager.

But start with the definitions. You might want to switch this pull request to a draft for the time being.

Reply:
I created custom classes and type-hinted all the way to rsemanager. 

Marked `prefix`  and `extended_attributes` of `RSEProtocolDict` with `#FIXME: typing.NotRequired`.  

Also it seems `_format_get_rse_protocols` needs to be type-hinted more precisely to have compatible return value type (mypy shows it is expected to return dict[str, Any]).

Reply:
Very well. I’ll have a closer look tomorrow, but already I noticed something.

`typing.NotRequired` should be used when a key may not exist in the dictionary. I have the impression that `prefix` always exists, is a string , but may be an empty string. `extended_attributes` is more complex. It always exists too, but its value may be `None`, `str` or some dictionary whose types are unclear.

At least that’s my understanding from reading `_format_get_rse_protocols()`. Have you found signs to the contrary?

Reply:
Oh, I thought that `Optional[]` and `nullable=True` in `models.RSEProtocols` could signify that key may not be passed to the dictionary. I'll check presence of these keys more carefully in this case.